### ISSUE TITLE
`--verbose` causes immediate exit

### ISSUE THREAD

**Issue by aphedges** [yt-dlp/yt-dlp#2659]
### Checklist

- [X] I'm reporting a bug unrelated to a specific site
- [X] I've verified that I'm running yt-dlp version **2022.02.04**. ([update instructions](https://github.com/yt-dlp/yt-dlp#update))
- [X] I've checked that all provided URLs are alive and playable in a browser
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/ytdl-org/youtube-dl#video-url-contains-an-ampersand-and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Description

Using the `--verbose` flag causes `yt-dlp` to closed immediately with an error code. Leaving out the flag allows it to work successfully:

```console
$ yt-dlp --version
2022.02.04
$ time yt-dlp --verbose https://www.youtube.com/watch?v=aetXqd9B8WE

real    0m0.329s
user    0m0.000s
sys     0m0.015s
$ echo $?
1
$ time yt-dlp https://www.youtube.com/watch?v=aetXqd9B8WE
[youtube] aetXqd9B8WE: Downloading webpage
[youtube] aetXqd9B8WE: Downloading android player API JSON
[info] aetXqd9B8WE: Downloading 1 format(s): 248+251
[download] Destination: ORIGINAL SONG MVQ- Calliope Mori x Gawr Gura x DECO_27 [aetXqd9B8WE].f248.webm
[download] 100% of 31.97MiB in 00:03
[download] Destination: ORIGINAL SONG MVQ- Calliope Mori x Gawr Gura x DECO_27 [aetXqd9B8WE].f251.webm
[download] 100% of 2.36MiB in 00:00
[Merger] Merging formats into "ORIGINAL SONG MVQ- Calliope Mori x Gawr Gura x DECO_27 [aetXqd9B8WE].webm"
Deleting original file ORIGINAL SONG MVQ- Calliope Mori x Gawr Gura x DECO_27 [aetXqd9B8WE].f248.webm (pass -k to keep)
Deleting original file ORIGINAL SONG MVQ- Calliope Mori x Gawr Gura x DECO_27 [aetXqd9B8WE].f251.webm (pass -k to keep)

real    0m5.351s
user    0m0.000s
sys     0m0.015s
$ echo $?
0
```

I found this problem while trying to debug why downloading a livestream now has no output to show that something is actually being downloaded. I don't know if that is related or not, but I thought it worth mentioning.

I can't find a flag that gives the environment information like `--verbose` does, but I am using Python 3.9.10 on x86_64 Windows 10 using the mintty terminal from Git for Windows 2.35.1(2). Please let me know if any other information is needed.

### Verbose log

N/a

**Comment by aphedges** [comment]
I figured out that directly invoking `yt-dlp` doesn't work but running it as a module does:

```verbose
$ python -m yt_dlp --verbose https://www.youtube.com/watch?v=aetXqd9B8WE
[debug] Command-line config: ['--verbose', 'https://www.youtube.com/watch?v=aetXqd9B8WE']
[debug] Encodings: locale cp1252, fs utf-8, out cp1252 (No ANSI), err cp1252 (No ANSI), pref cp1252
[debug] yt-dlp version 2022.02.04 [c1653e9ef] (source)
[debug] Python version 3.9.10 (CPython 64bit) - Windows-10-10.0.19044-SP0
[debug] exe versions: ffmpeg 4.4.1-full_build-www.gyan.dev (setts), ffprobe 4.4.1-full_build-www.gyan.dev
[debug] Optional libraries: Cryptodome, mutagen, sqlite, websockets
[debug] Proxy map: {}
...
```

I still have no clue why the entry point fails with the `--verbose` flag but works with the `--quiet` flag or no flags at all.

**Comment by dirkf** [comment]
What happens if you use `-v` instead?

**Comment by aphedges** [comment]
It's the same failure.

I did some more experimentation and research after I created this issue last night, and I think it has to do with my terminal setup. The command works in Command Prompt but not mintty. Most of the encodings listed by the debug output in my previous comment are `cp1252`, even though my locale is set to `C.UTF-8`. In addition, I managed to figure out that stack traces are being swallowed, but I can force them to display with [`traceback.format_exc()`](http://docs.python.org/3/library/traceback.html#traceback.format_exc). It seems to be a printing issue. I think I have enough information to look into this more myself. I will either figure out a solution and create a PR for it or not figure out a solution but post here what I have found.

By the way, I'd like to note that I'm running `yt-dlp` on my personal Windows machine, but almost all my dev experience is on Unix systems. Windows Python seems to have very different handling of encodings, entry points, etc. compared to Unix Python, so a lot of things that I have expected to work a specific way don't. This will at least be a good learning experience for me.

**Comment by pukkandan** [comment]
I am currently not a position to set up mintty and test this. Can you try disabling this line? https://github.com/yt-dlp/yt-dlp/blob/a3125791c7a5cdf2c8c025b99788bf686edd1a8a/yt_dlp/YoutubeDL.py#L550

Otherwise, it'll take me a day or to to be able to test this properly

**Comment by aphedges** [comment]
I tried disabling the line and tried `youtube-dl`, and neither of them worked.

According to [bpo-45538](https://bugs.python.org/issue45538), MinGW is not a supported platform for Python. I expect that using the official Windows installation from Python.org through MinGW and mintty likely causes even more problems than using either the official installer with Command Prompt or MinGW Python with mintty. I shouldn't expect my setup to actually work properly as if I were using Python on a supported Windows or Unix setup. I did find a fix to all my problems by prepending `winpty` to my command. `winpty yt-dlp --verbose https://www.youtube.com/watch?v=aetXqd9B8WE` works without any problems. Another solution was to redirect either stdout or stderr to `/dev/null` when calling the command, but that obviously prevents some of the output from being displayed. Any future MinGW users trying to use `yt-dlp` should use one of these solutions to avoid any of the issues that I encountered. For anyone curious about what was actually going on, read after the horizontal line.

@dirkf and @pukkandan, thank you for your suggestions. I'm not sure there is anything you can do to solve this. I don't know if it's even possible to detect my particular setup to do an automatic warning to use `winpty`. The only possible mitigation I can think of is to detect early on if `sys.stderr is None` but `sys.stdout is not None`, but I'm not even sure what to do with that knowledge. Maybe fail with an error or print everything to stdout in that case? Do you have any ideas, or should I just close this issue?

---

I did a lot of print-line debugging by modifying the source in `site-packages`, and my first (successful) step was to get any visible error besides the `1` exit code. I added a general `except` block to [`__init__.py:main()`](https://github.com/yt-dlp/yt-dlp/blob/2022.02.04/yt_dlp/__init__.py#L855) with logic to print the stack trace to stdout. I got the following (paths shortened for readability):

```python
Traceback (most recent call last):
  File "Python39\lib\site-packages\yt_dlp\__init__.py", line 857, in main
    _real_main(argv)
  File "Python39\lib\site-packages\yt_dlp\__init__.py", line 75, in _real_main
    parser, opts, args = parseOpts(argv)
  File "Python39\lib\site-packages\yt_dlp\options.py", line 116, in parseOpts
    write_string(f'\n{root}'.replace('\n| ', '\n[debug] ')[1:] + '\n')
  File "Python39\lib\site-packages\yt_dlp\utils.py", line 2062, in write_string
    out.write(s)
AttributeError: 'NoneType' object has no attribute 'write'
```

For some weird reason, `sys.stderr` was `None`. The `[debug]` lines at the start of `--verbose` output are sent to stderr, so `yt-dlp` was crashing when it tried to write to `None`. No stack trace was being displayed because stack traces are normally printed to stderr, but there was no stderr to print to. After I ascertained that `sys.stderr` was `None` even at the very start of `__init__.py` (i.e., verified that it was not caused by something internal to the library), I created an extremely simply example package made from three files:

- `sys_test.py`:

  ```python
  import sys
  
  def main():
      print(sys.stdout)
      print(sys.stderr)
  
  if __name__ == "__main__":
      main()
  ```
- `setup.cfg`:

  ```ini
  [metadata]
  name = sys_test
  
  [options]
  py_modules =
          sys_test
  scripts =
          sys_test.py
  
  [options.entry_points]
  console_scripts =
          sys_test = sys_test:main
  ```
- `setup.py`:
  ```python
  from setuptools import setup
  setup()
  ```

I was able to successfully able to run the program as both a script and as an installed module:

```console
$ python sys_test.py
<_io.TextIOWrapper name='<stdout>' mode='w' encoding='cp1252'>
<_io.TextIOWrapper name='<stderr>' mode='w' encoding='cp1252'>
$ pip install .
...
Successfully installed sys-test-0.0.0
$ cd ..
$ python -m sys_test
<_io.TextIOWrapper name='<stdout>' mode='w' encoding='cp1252'>
<_io.TextIOWrapper name='<stderr>' mode='w' encoding='cp1252'>
```

However, using the entry point without `winpty` meant that `sys.stderr` was `None`:

```console
$ sys_test
<_io.TextIOWrapper name='<stdout>' mode='w' encoding='cp1252'>
None
$ winpty sys_test
<_io.TextIOWrapper name='<stdout>' mode='w' encoding='utf-8'>
<_io.TextIOWrapper name='<stderr>' mode='w' encoding='utf-8'>
```

I still don't know why this suddenly started being a problem for me when `yt-dlp` worked without any issues a couple of weeks ago. Reverting to a version from November still failed. However, the 2022-01-24 release of [Git for Windows v2.35.0](https://github.com/git-for-windows/build-extra/blob/main/ReleaseNotes.md#changes-since-git-for-windows-v2341-november-25th-2021) updated MinGW and mintty for the first time since the 2020-10-19 release of [Git for Windows v2.29.0](https://github.com/git-for-windows/build-extra/blob/main/ReleaseNotes.md#changes-since-git-for-windows-v2280-july-28th-2020). I therefore believe the regression actually lies in my setup than in any change in `yt-dlp`.

**Comment by pukkandan** [comment]
I suggest trying to report this to mingw devs first. If it is indeed something they can fix, there is no point in us trying to work around it. If they can't (or refuse to) fix this, we can add the workaround of writing to stdout when `stderr is None`

**Comment by dirkf** [comment]
Maybe s/t on the lines of `sys.stderr = open(2, 'w') if sys.stderr is None ...`? Or perhaps that will fail for the same reason that stderr isn't being set already.

Also, the manual (3.10.2/sys) says (this also suggests to check the value of `sys.__stderr__`):
> Note
>
> Under some conditions `stdin`, `stdout` and `stderr` as well as the original values `__stdin__`, `__stdout__` and `__stderr__` can be None. It is usually the case for Windows GUI apps that aren’t connected to a console and Python apps started with pythonw.

**Comment by aphedges** [comment]
I finally found the time to investigate this further. I tried downgrading Git for Windows to [v2.34.1](https://github.com/git-for-windows/git/releases/tag/v2.34.1.windows.1), and it didn't help. I decided to try downgrading `pip` and `setuptools` (because they install the entry points) and figured out the regression was in <https://github.com/pypa/pip/compare/21.3.1...22.0>. With the help of `git bisect`, I pinpointed the regression as being caused by <https://github.com/pypa/pip/commit/66c9d12317c0bcca8ad2930f2741aafa34c017c3>, which updated the vendored version of `distlib` to 0.3.4. I then checked <https://github.com/pypa/distlib/compare/0.3.3...0.3.4> and figured out that commit <https://github.com/pypa/distlib/commit/277cc4eb2eebbca4742004d1fcc6c08e1c70c3b9> was the core cause. It's hard to tell why it's happening because of no source code changes were committed to the repository. At the very least, figuring out where the issue was allowed me to find <https://github.com/pypa/pip/issues/10444> and <https://github.com/pypa/pip/issues/10875>, which discuss this bug. They pointed me to <https://bitbucket.org/vinay.sajip/simple_launcher/commits/4c263ce78a4b6b30274d877780e9525ac2f5c15e>, which is the likely source code change that caused the regression. A complete fix for the underlying bug might make it into `pip`, but it is likely to take a couple of months. Fortunately, this error is severe enough that according to <https://github.com/pypa/pip/issues/10875#issuecomment-1050790275> and <https://github.com/pypa/pip/milestone/65>, there will likely be a patch release soon to downgrade `distlib` to the previous working version. For a lengthy analysis of the bug's cause, I found <https://github.com/microsoft/vscode-python/issues/18561#issuecomment-1054306197>, but it doesn't really help us here.

I don't know what solution there should be in `yt-dlp`. @dirkf's suggestion might work:

```console
$ python -c 'import sys; sys.stderr = None; sys.stderr.write("foo")'
$ python -c 'import sys; sys.stderr = None; sys.stderr = open(2, "w"); sys.stderr.write("foo")'
foo
```

However, I'm not sure if this behavior will still work when using an entry point or how much effort should be put into a solution that is only needed for a `pip` bug that will last only a couple of months. For now, I'm just running an old version of `pip` where this bug does not exist.

**Comment by dirkf** [comment]
Once upon a time MingGW was about the only open source example for shimming Win32 into POSIX. Now everybody can get it wrong ...

**Comment by pukkandan** [comment]
> I suggest trying to report this to mingw devs first

Any update on this?

**Comment by aphedges** [comment]
Sorry, but I forgot to update this issue. A `pip` PR  (<https://github.com/pypa/pip/pull/10943>) downgraded their vendored `distlib` version to 0.3.3, and it was released as part of `pip` [22.0.4](https://github.com/pypa/pip/releases/tag/22.0.4), fixing the issue.

> > I suggest trying to report this to mingw devs first
> 
> Any update on this?

I never ended up reporting the problem to the mingw devs because my investigation in <https://github.com/yt-dlp/yt-dlp/issues/2659#issuecomment-1057202696> showed it was a `pip` problem.

Unless you think it's worth creating a workaround for missing `stderr`, I believe this issue should be closed.

**Comment by pukkandan** [comment]
Since pip has addressed the issue, I don't think it is worth creating a workaround. If we find some other instance of missing stderr, we can have a second look at this
